File: legacy/py-prototype/toposort.py
from __future__ import annotations

from collections import defaultdict, deque


class CycleError(ValueError):
	pass


def _name_of(item: dict) -> str:
	return item.get("name") or f"{item['repo']}@{item['branch']}"


def levels(items: list[dict]) -> list[list[str]]:
	"""Return deterministic topo levels of item names.
	Raises CycleError if a cycle exists.
	Raises KeyError if a dep references a missing item.
	"""
	names = [_name_of(i) for i in items]
	name_set = set(names)
	# Build graph
	in_deg: dict[str, int] = {n: 0 for n in names}
	children: dict[str, list[str]] = defaultdict(list)
	for it in items:
		me = _name_of(it)
		for d in it.get("deps", []) or []:
			if d not in name_set:
				raise KeyError(f"unknown dep '{d}' for item '{me}'")
			children[d].append(me)
			in_deg[me] += 1
	# Kahn's algorithm with deterministic ordering
	q = deque(sorted([n for n, deg in in_deg.items() if deg == 0]))
	result: list[list[str]] = []
	while q:
		this_level = sorted(list(q))
		q.clear()
		result.append(this_level)
		for n in this_level:
			for c in sorted(children.get(n, [])):
				in_deg[c] -= 1
				if in_deg[c] == 0:
					q.append(c)
	# If any node still has in-degree, cycle exists
	if any(deg > 0 for deg in in_deg.values()):
		raise CycleError("dependency cycle detected")
	return result

File: src/core/toposort.ts
import { PlanItem } from "./plan.js";

export class CycleError extends Error {
	constructor(message: string = "dependency cycle detected") {
		super(message);
		this.name = "CycleError";
	}
}

function nameOf(item: PlanItem): string {
	return item.branch; // Using branch as the identifier for now
}

export function computeLevels(items: PlanItem[]): number[][] {
	/**
	 * Return deterministic topo levels of item indices.
	 * Throws CycleError if a cycle exists.
	 * Throws Error if a dependency references a missing item.
	 */
	const itemMap = new Map<number, PlanItem>();
	const nameToId = new Map<string, number>();
	
	// Build maps
	for (const item of items) {
		itemMap.set(item.id, item);
		nameToId.set(nameOf(item), item.id);
	}
	
	// Build graph using item IDs
	const inDegree = new Map<number, number>();
	const children = new Map<number, number[]>();
	
	// Initialize
	for (const item of items) {
		inDegree.set(item.id, 0);
		children.set(item.id, []);
	}
	
	// Build dependencies
	for (const item of items) {
		for (const depId of item.needs) {
			if (!itemMap.has(depId)) {
				throw new Error(`unknown dependency ${depId} for item ${item.id}`);
			}
			children.get(depId)!.push(item.id);
			inDegree.set(item.id, inDegree.get(item.id)! + 1);
		}
	}
	
	// Kahn's algorithm with deterministic ordering
	let queue = Array.from(inDegree.entries())
		.filter(([, deg]) => deg === 0)
		.map(([id]) => id)
		.sort();
	
	const result: number[][] = [];
	
	while (queue.length > 0) {
		const thisLevel = [...queue].sort();
		queue = [];
		result.push(thisLevel);
		
		for (const id of thisLevel) {
			const childList = children.get(id) || [];
			for (const childId of childList.sort()) {
				const newDegree = inDegree.get(childId)! - 1;
				inDegree.set(childId, newDegree);
				if (newDegree === 0) {
					queue.push(childId);
				}
			}
		}
	}
	
	// Check for cycles
	if (Array.from(inDegree.values()).some(deg => deg > 0)) {
		throw new CycleError();
	}
	
	return result;
}